api/src/main/java/org/openmrs/api/impl/ProgramWorkflowServiceImpl.java

```diff
@@ -74,34 +74,39 @@ public Program saveProgram(Program program) throws APIException {
 			throw new APIException("Program.concept.required", (Object[]) null);
 		}
 		
-		// ProgramWorkflow
 		for (ProgramWorkflow workflow : program.getAllWorkflows()) {
-			
-			if (workflow.getConcept() == null) {
-				throw new APIException("ProgramWorkflow.concept.required", (Object[]) null);
-			}
-			if (workflow.getProgram() == null) {
-				workflow.setProgram(program);
-			} else if (!workflow.getProgram().equals(program)) {
+			ensureProgramIsSet(workflow, program);
+			if (!workflow.getProgram().equals(program)) {
 				throw new APIException("Program.error.contains.ProgramWorkflow", new Object[] { workflow.getProgram() });
 			}
-			
-			// ProgramWorkflowState
 			for (ProgramWorkflowState state : workflow.getStates()) {
-				
-				if (state.getConcept() == null || state.getInitial() == null || state.getTerminal() == null) {
-					throw new APIException("ProgramWorkflowState.requires", (Object[]) null);
-				}
-				if (state.getProgramWorkflow() == null) {
-					state.setProgramWorkflow(workflow);
-				} else if (!state.getProgramWorkflow().equals(workflow)) {
-					throw new APIException("ProgramWorkflow.error.contains.state", new Object[] { workflow.getProgram() });
-				}
+				ensureProgramWorkflowIsSet(state, workflow);
 			}
 		}
 		return dao.saveProgram(program);
 	}
 	
+	private void ensureProgramIsSet(ProgramWorkflow workflow, Program program) throws APIException {
+		if (workflow.getConcept() == null) {
+			throw new APIException("ProgramWorkflow.concept.required", (Object[]) null);
+		}
+		if (workflow.getProgram() == null) {
+			workflow.setProgram(program);
+		}
+	}
+	
+	private void ensureProgramWorkflowIsSet(ProgramWorkflowState state, ProgramWorkflow workflow) throws APIException {
```

Comment:
same here remove the `throws APIException`, other than that this method is perfect 👍